src/runtime-helpers/runtime-helpers.ts

```diff
@@ -213,6 +227,37 @@ module ScriptLab {
     function _isDialogApiSupported() {
         return (window as any).Office.context.requirements.isSetSupported('DialogApi');
     }
+
+    function _extractAndCacheAccessToken(message: string, clientId: string, resource: string): string {
+        const authResponseKeyValues = message.split('&');
+        const accessToken = authResponseKeyValues[0].substr('AUTH:access_token='.length);
+        const expiresIn = Number(authResponseKeyValues[1].substr('AUTH:expires_in='.length));
+        cachedAuthTokens[clientId + resource] = { token: accessToken, expiry: Date.now() + ((expiresIn - 60) * 1000) };
+        return accessToken;
+    }
+
+    function _getCachedAccessToken(clientId: string, resource: string): string {
+        const cachedAuthToken = cachedAuthTokens[clientId + resource];
+        if (cachedAuthToken) {
+            if (cachedAuthToken.expiry > Date.now()) {
+                /* Token is valid. Return it */
+                return cachedAuthToken.token;
+            }
+            else {
+                /* Token in cache has expired. Remove it from the cache */
+                _removeCachedAccessToken(clientId, resource);
+                return null;
+            }
+        } else {
+            return null;
+        }
+    }
+
+    function _removeCachedAccessToken(clientId: string, resource: string) {
+        if (cachedAuthTokens[clientId + resource]) {
```

Comment:
Nit: do you want to create a helper method for concatenating.  I realize it seems silly, but what if you later want to add something else, or change the contactenation to have an underscore in-between (I would...).  Then you'll end up chasing the same logic all throughout the code, instead of changing only one function.